lib/DependencyProvider.js

```diff
@@ -0,0 +1,317 @@
+// @flow
+
+const fs = require('fs');
+const os = require('os');
+const path = require('path');
+const request = require('sync-request').default;
+
+// Cache of existing versions according to the package website.
+let onlineVersionsCache /*: Map<string, Array<string>> */ = new Map();
+
+// Memoization cache to avoid doing the same work twice in listAvailableVersions.
+// This is to be cleared before each call to solve_deps().
+const listVersionsMemoCache /*: Map<string, Array<string>> */ = new Map();
+
+function fetchElmJsonOnline(
+  pkg /*: string */,
+  version /*: string */
+) /*: string */ {
+  try {
+    return fetchElmJsonOffline(pkg, version);
+  } catch (_) {
+    const remoteUrl = remoteElmJsonUrl(pkg, version);
+    const elmJson = request('GET', remoteUrl).getBody('utf8'); // need utf8 to convert from gunzipped buffer
+    const cachePath = cacheElmJsonPath(pkg, version);
+    const parentDir = path.dirname(cachePath);
+    fs.mkdirSync(parentDir, { recursive: true });
+    fs.writeFileSync(cachePath, elmJson);
+    return elmJson;
+  }
+}
+
+function fetchElmJsonOffline(
+  pkg /*: string */,
+  version /*: string */
+) /*: string */ {
+  console.log('Fetching: ' + pkg + ' @ ' + version);
```

Comment:
Are these `console.log`s temporary?

If we want to log something, I think it should be part of the `Compiling > Solving dependencies > Starting tests` “progress bar”. I noticed “Solving dependencies” has been removed. But this approach still makes network requests, right? If so, isn’t “Solving dependencies” still needed since it can take time?

Reply:
Yes the logs are temporary, definitely better to have proper logging. I think I removed the logging because I wanted to remove the logging related to downloading elm-json, I don't remember for sure but I had the impression that the logging logic for solving dependencies and for downloading elm-json was tied so I removed all of it.

In practice, with the few caches, the solver should almost never make http requests. Typically, there will be some requests the first time a package is seen by the solver on the user machine. Then, the cache in `~/.elm/pubgrub/` will be enough to solve everything offline, even in other project folders since the cache does not live in `elm-stuff/`. But it's probably better to re-add a `Solving dependencies` step?